addons/controllers/calico/calicoconfig_controller.go

```diff
@@ -6,43 +6,220 @@ package controllers
-// the user.
 func (r *CalicoConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("calicoconfig", req.NamespacedName)
+	r.Log = r.Log.WithValues("CalicoConfig", req.NamespacedName)
+
+	r.Log.Info("Start reconciliation")
+
+	// fetch CalicoConfig resource, ignore not-found errors
+	calicoConfig := &cniv1alpha1.CalicoConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, calicoConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("CalicoConfig resource not found")
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "Unable to fetch CalicoConfig resource")
+		return ctrl.Result{}, err
+	}
+
+	// deep copy CalicoConfig to avoid issues if in the future other controllers where interacting with the same copy
+	calicoConfig = calicoConfig.DeepCopy()
 
-	// TODO(user): your logic here
+	// config resources are expected to have the same name as the cluster. However, we ideally try to read the cluster name from the owner reference of the addon config object
+	clusterName := req.NamespacedName.Name
+	for _, ownerRef := range calicoConfig.GetOwnerReferences() {
+		if ownerRef.Kind == "Cluster" {
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+
+	// verify that the cluster related to config is present
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: clusterName}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' not found", req.Namespace, clusterName))
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, fmt.Sprintf("Unable to fetch cluster '%s/%s'", req.Namespace, clusterName))
+		return ctrl.Result{}, err
+	}
+
+	// reconcile CalicoConfig resource
+	if retResult, err := r.ReconcileCalicoConfig(calicoConfig, cluster); err != nil {
+		r.Log.Error(err, "Unable to reconcile CalicoConfig")
+		return retResult, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *CalicoConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *CalicoConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	r.Ctx = ctx
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&cniv1alpha1.CalicoConfig{}).
+		WithOptions(options).
 		Complete(r)
 }
+
+// ReconcileCalicoConfig reconciles CalicoConfig CR
+func (r *CalicoConfigReconciler) ReconcileCalicoConfig(
+	calicoConfig *cniv1alpha1.CalicoConfig,
+	cluster *clusterapiv1beta1.Cluster) (_ ctrl.Result, retErr error) {
+
+	var patchConfig bool
+
+	patchHelper, err := clusterapipatchutil.NewHelper(calicoConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// patch CalicoConfig before returning the function
+	defer func() {
+		// patchConfig will be true if owner reference or secretRef are added
+		if patchConfig {
```

Comment:
Don't need separate bool to check if it was changed, patchHelper already should do the diff checking

Reply:
cc: @blc1996